test/test_modules.py

```diff
@@ -156,6 +156,95 @@ def test_pickle(self, device, dtype, module_info):
                     output_from_copy = m_copy(*args, **kwargs)
                     self.assertEqual(output, output_from_copy)
 
+    @modules(module_db, allow_devices={"cuda"})
+    def test_cpu_gpu_results_are_equal(self, device, dtype, module_info):
+        # Test cpu and gpu results are the same
+        module_cls = module_info.module_cls
+        module_inputs_cpu = module_info.module_inputs_func(module_info, device="cpu", dtype=dtype,
+                                                           requires_grad=True)
+
+        def make_leafs_input(items):
```

Comment:
This is for cases where the input are not leaves. For example in NLLLoss, the forward_input args are not leaves:

```python
import torch
from torch.testing import make_tensor

input = make_tensor((15, 10), device='cpu', dtype=torch.float32, requires_grad=True).log_softmax(dim=1)
input.is_leaf
# False
```

Reply:
WHy do you need them to be leafs?
And the only way for them not to be leafs is if they do require gradients so not a problem for gradcheck

Reply:
This PR is checking for cpu+gpu parity for `_.grad`. `backward()` only accumulates gradients in the leafs:

```python
import torch
from torch.testing import make_tensor
import torch.nn as nn

input = (make_tensor((15, 10), device='cpu', dtype=torch.float32, requires_grad=True)
         .log_softmax(dim=1))
print("input.requires_grad:", input.requires_grad)

target = torch.empty(15, device='cpu').uniform_().mul(10).floor().long()

nll = nn.NLLLoss()
loss = nll(input, target)
loss.backward()
print("input.grad:", input.grad)

# input.requires_grad: True
# input.grad: None
```

This also issues a `UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed.`

Reply:
In that case, 
```python
for item in items:
    if isinstance(item, torch.Tensor):
        item.retain_grad()
```